File: Algorithm/2023.10/graywhales/입국심사.java
/*
  시간을 binarySearch 탐사대상으로 두고
  조건에 부합할때 right를 mid로 부합하지 않을때 left를 mid+1로 옮겨가면서 탐색
  초기 right는 min*n으로 설정
*/
class Solution {
    public long solution(int n, int[] times) {
        int min=Integer.MAX_VALUE;
        for(int i=0;i<times.length;i++){
            min=Math.min(min,times[i]);
        }
        long left=1;
        long right=(long)min*n; //min*n integer범위임을 주의
        long mid;
        long sum;
        while(left<right){
            mid=(left+right)/2;
            sum=0;
            for(int i=0;i<times.length;i++){
                sum+=mid/times[i];
            }
            if(sum>=n){
                right=mid;
            }else{
                left=mid+1;
            }
        }
        return right;
    }
}

File: Algorithm/2023.10/Wolfy916/알고리즘 테마정리/이분탐색/입국심사.py
'''
제목: 입국심사 - 프로그래머스(레벨3)
분류: 이분탐색
1. 이분탐색을 통해 mid분내에 모든 심사가 완료되는지 체크
2. 최적해 탐색
'''

def solution(n, times):
    def check(m):
        nonlocal n
        cnt = 0
        for time in times:
            cnt += m // time
            if cnt >= n: return True
        return False
    
    answer = 0
    left, right = 1, times[-1] * n
    while left <= right:
        mid = (left + right) // 2
        if check(mid):
            answer = mid
            right = mid - 1
        else:
            left = mid + 1

    return answer